In [9]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')

In [10]:
trade_history = pd.read_csv('data/processed.csv')

In [11]:
x = trade_history.groupby(['Coin','classification']).agg({
    'Closed PnL': ['sum','mean'],
    'Trade ID': 'count'
})

In [12]:
x.columns = ['total_pnl', 'avg_pnl', 'trade_count']

x = x.reset_index()

## Top Performing Coins Based on Sentiment and PnL

In [13]:

top_profit = x.sort_values(
    'trade_count',
    ascending=False
)

top_5_best = (
    top_profit
    .sort_values(['classification', 'avg_pnl'], ascending=[True, False])
    .groupby('classification')
    .head(5)
)

top_5_best['rank'] = (
    top_5_best.groupby('classification')['avg_pnl']
    .rank(method='first', ascending=False)
)

top_5_best

,Coin,classification,total_pnl,avg_pnl,trade_count,rank
523,PNUT,Extreme Fear,1.855649e+03,1855.649224,1,1.0
449,MELANIA,Extreme Fear,9.737917e+04,218.338940,446,2.0
310,ETH,Extreme Fear,2.743202e+05,196.927660,1393,3.0
599,SUI,Extreme Fear,4.478899e+04,179.875468,249,4.0
626,TRUMP,Extreme Fear,3.163804e+04,174.795818,181,5.0
519,PEOPLE,Extreme Greed,3.606460e+02,360.646005,1,1.0
155,@85,Extreme Greed,1.896824e+04,291.819153,65,2.0
279,CYBER,Extreme Greed,1.915296e+02,191.529600,1,3.0
10,@107,Extreme Greed,1.988619e+06,191.158211,10403,4.0
468,MOODENG,Extreme Greed,1.608388e+04,165.813219,97,5.0


## Key Insights: Top Performing Coins by Sentiment Regime

### Extreme Fear
- `PNUT` achieved the highest avg pnl:
  - **$1,855.65 avg pnl** from only **1 trade**
- Among high-volume assets:
  - `ETH` generated:
    - **$196.93 avg pnl**
    - across **1,393 trades**
- `MELANIA` showed strong consistency:
  - **446 trades**
  - **$218.34 avg pnl**

### Extreme Greed
- `@107` dominated profitability at scale:
  - **$1.99M total pnl**
  - **10,403 trades**
  - **$191.16 avg pnl**
- `MOODENG` maintained high profitability:
  - **$165.81 avg pnl**
- Several top performers had very low trade counts, indicating potential outlier-driven profitability.

### Fear
- `SEI` achieved the highest avg pnl:
  - **$835.60 avg pnl**
- `ENA` combined both profitability and meaningful activity:
  - **193 trades**
  - **$830.65 avg pnl**
- `DOGE` and `ARB` also showed strong contrarian profitability during fearful conditions.

### Greed
- `SOL` was the strongest large-scale performer:
  - **$451.7K total pnl**
  - **1,586 trades**
  - **$284.80 avg pnl**
- `AVAX` and `ETHFI` also showed strong profitability during bullish sentiment regimes.

### Neutral
- `DOGE` generated the highest avg pnl overall:
  - **$2,969.47 avg pnl**
- `AVAX` and `ZRO` also showed extremely high profitability:
  - `AVAX` → **$1,610.91 avg pnl**
  - `ZRO` → **$1,446.90 avg pnl**
- `SOL` maintained stable profitability at scale:
  - **1,400 trades**
  - **$216.70 avg pnl**

- Large-cap assets such as `ETH` and `SOL` demonstrated more stable profitability across multiple sentiment regimes.
- Several smaller-cap or meme-related assets exhibited extremely high avg pnl with very low trade counts, indicating potential outlier effects and elevated volatility.

In [14]:
# Top 5 worst coins for each sentiment classification
top_5_worst = (
    top_profit
    .sort_values(['classification', 'avg_pnl'], ascending=[True, True])
    .groupby('classification')
    .head(5)
)

top_5_worst['rank'] = (
    top_5_worst.groupby('classification')['avg_pnl']
    .rank(method='first', ascending=True)
)

top_5_worst

,Coin,classification,total_pnl,avg_pnl,trade_count,rank
319,FARTCOIN,Extreme Fear,-138014.106207,-244.272754,565,1.0
170,ADA,Extreme Fear,-31317.986798,-182.081319,172,2.0
568,S,Extreme Fear,-11474.720000,-149.022338,77,3.0
361,GRIFFAIN,Extreme Fear,-378.092427,-94.523107,4,4.0
9,@107,Extreme Fear,-136085.368300,-89.588788,1519,5.0
36,@135,Extreme Greed,-5981.363990,-5981.363990,1,1.0
294,DYM,Extreme Greed,-255.946851,-127.973425,2,2.0
450,MELANIA,Extreme Greed,-18726.171999,-70.399143,266,3.0
369,HPOS,Extreme Greed,-882.628857,-33.947264,26,4.0
160,@9,Extreme Greed,-58.564889,-29.282445,2,5.0


# Key Insights: Worst Performing Coins by Sentiment Regime

## Extreme Fear
- `FARTCOIN` recorded the largest losses:
  - **-$244.27 avg pnl**
  - across **565 trades**
- `@107` generated substantial cumulative losses:
  - **-$136.1K total pnl**
  - over **1,519 trades**
- `ADA` also showed heavy underperformance:
  - **-$182.08 avg pnl**

## Extreme Greed
- `@135` recorded the worst avg pnl overall:
  - **-$5,981.36 avg pnl**
  - from a single trade
- `MELANIA` produced significant losses at scale:
  - **266 trades**
  - **-$70.40 avg pnl**

## Fear
- `WIF` showed the worst profitability:
  - **-$237.63 avg pnl**
- `KAITO` generated large cumulative losses:
  - **-$14.9K total pnl**
  - across **153 trades**
- `FARTCOIN` remained consistently loss-making:
  - **951 trades**
  - **-$11.37 avg pnl**

## Greed
- `TRUMP` exhibited the most severe losses across all regimes:
  - **-$848.85 avg pnl**
  - **-$432.9K total pnl**
  - across **510 trades**
- `IO` also showed substantial negative performance:
  - **-$77.39 avg pnl**
  - across **281 trades**

## Neutral
- `BERA` generated the largest neutral-market losses:
  - **-$189.62 avg pnl**
  - across **114 trades**
- `AAVE` remained consistently unprofitable:
  - **-$106.19 avg pnl**

- Meme and speculative assets showed the highest downside volatility across emotional market regimes.
- Several extreme losses occurred on low trade counts, indicating strong outlier influence and elevated risk concentration.
- `TRUMP`, `FARTCOIN`, and `MELANIA` consistently underperformed during emotionally charged market conditions.